#28 Remove illegal characters from caching hash according to PSR6 specs

samvdb opened the pull request:
See #28

samvdb committed:
Remove illegal characters from caching hash according to PSR6 specs

Bilge commented on src/Connector/CachingConnector.php:
```diff
@@ -76,6 +76,6 @@ public function isCacheEnabled()
     private function hash(array $structure)
     {
-        return json_encode($structure);
+        return preg_replace('/[\{\}\(\)\/\\\@:]/', '.', json_encode($structure));
```
We should prefer `str_replace()` here because regular expressions must be compiled and are therefore typically slower to execute. Consider assigning the array of replacement characters to a local static variable to lazily allocate the replacements list once instead of for each method invocation.

Bilge replied:
Was there any particular reason for choosing the dot (`.`) as a replacement character?

Bilge reviewed (changes requested):
Hi @samvdb :wave: Thanks for the PR :+1: 

You are correct that we must respect this rule in order to be PSR-6 compliant. However, we need some tests to confirm this fix is working correctly.

samvdb commented:
Well the dot is a valid character according to the RFC so its a sure bet. However i replaced the regex with strreplace for performance and removed the dot. This can cause empty hashes in some cases i think. A better approach would be hashing the json_encode string, however this might impact performance. What would you recommend?

Bilge commented:
It seems there are advantages and disadvantages for each approach.

### Hashing 
* :white_check_mark:  Key length constrained to reasonable limit with fixed length.
* :negative_squared_cross_mark: Slight performance hit.

Performance hit could be mitigated by selecting an appropriate hash implementation. We do not need a cryptographic hash because security is not a concern; we just need a hash with near-guaranteed uniqueness for different input data.

### String replacement
* :white_check_mark: Faster.
* :negative_squared_cross_mark: Variable key length with no upper limit.
* :negative_squared_cross_mark: Increased chance of collision due to malformed JSON (to be verified).

Which do you prefer?

samvdb commented:
Another con for hashing is that debugging will become harder because keys are not readable anymore. I will stick with str_replace

Bilge commented:
Very well, however consider we may still switch to hashing in future if the need arises.

For now, let's replace the banned characters with a single character to avoid the chance of collisions (instead of removing them). It is important that it is not possible to deliberately craft a request to create the same key for different data, which should be easier to avoid if we do replacement instead of removal.

samvdb committed:
Add unit test for PSR6 reserved characters

samvdb commented:
I have added tests and will add the dot again.
Would you mind if i changed the hash function to protected instead of private?

samvdb committed:
Add dot character as replacement

Bilge commented:
I do not think the hash function should be `protected` because I do not see the need for inheritors to override it. I presume you are asking because you find it difficult to test as a private method?

samvdb commented:
No i was asking because other connectors might want to extend cachingconnector and implement hashing if need be. 

I'll leave it private since the issue should be resolved for the moment.

Bilge commented on test/Unit/Porter/Net/Caching/CachingConnectorTest.php:
```diff
@@ -0,0 +1,29 @@
+{
+    use MockeryPHPUnitIntegration;
+
+    public function test()
```
* I would expect this test case to be added to the existing [CachingConnectorTest](https://github.com/ScriptFUSION/Porter/blob/master/test/Integration/Porter/Connector/CachingConnectorTest.php)
* This test case should have a name, e.g. *testHashExcludesBannedCharacters*.

Bilge commented on test/Unit/Porter/Net/Caching/CachingConnectorTest.php:
```diff
@@ -0,0 +1,29 @@
+
+final class CachingConnectorTest extends \PHPUnit_Framework_TestCase
+{
+    use MockeryPHPUnitIntegration;
```
It seems we're not using Mockery here (although perhaps we should be?) As is, this can be removed.

Bilge commented on test/Unit/Porter/Net/Caching/CachingConnectorTest.php:
```diff
@@ -0,0 +1,29 @@
+
+    public function test()
+    {
+        $memoryCache = new MemoryCache();
```
For this to quality as a unit test it can only instantiate the system under test (SUT). Either mock the cache or move this to the integration test class.

Bilge commented on test/Unit/Porter/Net/Caching/CachingConnectorTest.php:
```diff
@@ -0,0 +1,29 @@
+        $structure = [$source, []];
+        self::assertSame('bar', $connector->fetch($source, null));
+
+        $hash = str_replace(str_split('{}()/\@:'), '.', json_encode($structure));
```
Seeing business logic duplicated in a test is usually an indication of a problem. In this case, the problem seems to be we're testing the exact output of the hashing function instead of just what we really care about: **the banned characters are excluded**. This creates a very brittle test because any change to the code causes the test to fail. 

Let's just test that the hash string doesn't contain the banned characters, so if the implementation changes, the test should continue to pass.

Bilge reviewed (changes requested):
I think the tests need some more work to be merged. However, I do not require you to make any changes if you do not want to. If you do not need this feature urgently I will look at it this evening and you should have a tag available by tomorrow. If you need this feature today, feel free to continue working on this.

Bilge commented:
This is now [merged into master](https://github.com/ScriptFUSION/Porter/compare/64037b0a9826566dcde4be2a3876047a4319e438...506e1c615af6f1389179b3677d7f75ed65c3a4ef) and tagged as 🏷 [3.2.1](https://github.com/ScriptFUSION/Porter/releases/tag/3.2.1).

Thanks @samvdb for your work 👍

samvdb commented:
Thanks Bilge!
